activerecord/lib/active_record/relation.rb

```diff
@@ -304,83 +304,157 @@ def many?
-    def cache_key(timestamp_column = :updated_at)
+    #    Product.where("name like ?", "%Game%").cache_key(timestamp_column: :last_reviewed_at)
+    #
+    # You can also specify a column other than a model's primary ID to generate
+    # the first & last ID segments of the cache_version for collections using a
+    # limit or offset.
+    #
+    #    Product.where("name like ?", "%Game%").cache_key(id_column: :upc)
+    def cache_key(deprecated_timestamp_column = nil, id_column: primary_key, timestamp_column: :updated_at)
+      if deprecated_timestamp_column
+        ActiveSupport::Deprecation.warn("Passing a timestamp column as a positional argument to ActiveRecord::Relation#cache_key is deprecated. Please use a `:timestamp_column` keyword argument instead.")
+        timestamp_column = deprecated_timestamp_column
+      end
+
       @cache_keys ||= {}
-      @cache_keys[timestamp_column] ||= klass.collection_cache_key(self, timestamp_column)
+      @cache_keys[[id_column, timestamp_column]] ||= klass.collection_cache_key(self, id_column: id_column, timestamp_column: timestamp_column)
     end
 
-    def compute_cache_key(timestamp_column = :updated_at) # :nodoc:
+    def compute_cache_key(id_column: primary_key, timestamp_column: :updated_at) # :nodoc:
       query_signature = ActiveSupport::Digest.hexdigest(to_sql)
       key = "#{klass.model_name.cache_key}/query-#{query_signature}"
 
-      if cache_version(timestamp_column)
+      if cache_version(id_column: id_column, timestamp_column: timestamp_column)
         key
       else
-        "#{key}-#{compute_cache_version(timestamp_column)}"
+        "#{key}-#{compute_cache_version(id_column: id_column, timestamp_column: timestamp_column)}"
       end
     end
     private :compute_cache_key
 
     # Returns a cache version that can be used together with the cache key to form
-    # a recyclable caching scheme. The cache version is built with the number of records
-    # matching the query, and the timestamp of the last updated record. When a new record
-    # comes to match the query, or any of the existing records is updated or deleted,
-    # the cache version changes.
+    # a recyclable caching scheme. The cache version is built with the number of records|
+    # matching the query, the IDs of the query's first and last records (if the query
+    # has a limit or offset and you're using MySQL or a version of PostgreSQL or SQLite that
+    # supports window functions), and the timestamp of the last updated record. When a
+    # new record comes to match the query, or any of the existing records is updated or
+    # deleted, the cache version changes.
     #
     # If the collection is loaded, the method will iterate through the records
     # to generate the timestamp, otherwise it will trigger one SQL query like:
     #
     #    SELECT COUNT(*), MAX("products"."updated_at") FROM "products" WHERE (name like '%Cosmic Encounter%')
```

Comment:
As we are one level up in complexity, adding an example query for both cases (PG/SQLlite vs. MySQL) would be nice for anyone to figure out what will happen (or finding this part of the doc starting from a query from its logs) 😃

Reply:
I'm a little hesitant to explain the underlying interface in too much detail, but the point about folks looking at a query from their logs and searching for more info seems reasonable. So, I've extended the documentation accordingly.

(Also, while my original solution used window functions, I later determined common table expressions were more efficient - I've corrected that as well.)